File: y22/m04/d25/binary_search.py
array = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]


def binary_search(array, target, start, end):
    mid = (start + end) // 2

    if start > end:
        return -1

    if array[mid] == target:
        return mid
    elif array[mid] > target:
        return binary_search(array, target, start, mid - 1)
    elif array[mid] < target:
        return binary_search(array, target, mid + 1, end)


print(binary_search(array, 4, 0, len(array) - 1))

File: y22/m12/d25/BinarySearch.java
package m12.d25;

public class BinarySearch {
    public static void main(String[] args) {
        Solution solution = new Solution();
//        int[] nums = {-1, 0, 3, 5, 9, 12};
//        int target = 9;
//        int correctAnswer = 4;

        int[] nums = {-1, 0, 5};
        int target = -1;
        int correctAnswer = 0;
        int answer = solution.search(nums, target);
        System.out.println(correctAnswer == answer);
        System.out.println(answer);
    }
}

class Solution {
    public int search(int[] nums, int target) {
        int left = 0;
        int right = nums.length - 1;
        while (left <= right) {
            int mid = (left + right) / 2;
            int now = nums[mid];
            if (now == target) {
                return mid;
            } else if (now > target) {
                right = mid - 1;
            } else left = mid + 1;
        }

        return -1;
    }
}
